File: 백준/Silver/15652. N과 M （4）/N과 M （4）.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class Main {
	public static void main(String[] args) throws IOException {
		input();
		solution();
	}
	private static int N, M;
	private static StringBuilder sb;
	private static void solution() {
		int[] arr = IntStream.rangeClosed(1, N).toArray();
		int[] res = new int[M];
		sb = new StringBuilder();
		comb(arr, res, 0, 0);
		System.out.println(sb.toString());
	}
	private static void comb(int[] arr, int[] res, int idx, int depth) {
		if (depth == M) {
			sb.append(
				Arrays.stream(res).mapToObj(String::valueOf).collect(Collectors.joining(" "))
			).append("\n");
			return;
		}

		for (int i = idx; i < N; i++) {
			res[depth] = arr[i];
			comb(arr, res, i, depth + 1);
		}
	}
	private static void input() throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int[] nm = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
		N = nm[0];
		M = nm[1];
	}
}

File: 백준/Silver/15652. N과 M （4）/N과 M （4）.kt
fun main() = with(System.`in`.bufferedReader()) {
    val (n, m) = readln().split(" ").map { it.toInt() }
    close()

    val arr = IntArray(n) { it + 1 }
    comb(arr, 0, 0, m)
    println(sb.toString())
}

private fun comb(arr : IntArray, depth : Int, idx : Int,  m : Int) {
    if (depth == m) {
        toString(arr, m)
        return
    }

    for (i in idx .. arr.lastIndex) {
        arr[depth] = i + 1
        comb(arr, depth + 1, i, m)
    }
}

private val sb = StringBuilder()

private fun toString(arr : IntArray, m : Int) =
    sb.append(
        arr.slice(0 until m)
            .joinToString(" "))
        .append("\n")